pkg/storegateway/bucket_index_reader.go

```diff
@@ -268,6 +224,84 @@ func (r *bucketIndexReader) expandedPostings(ctx context.Context, ms []*labels.M
 	return ps, nil
 }
 
+// toPostingGroups returns a set of labels for which to look up postings lists. It guarantees that
+// each postingGroup's keys exist in the index. The order of the returned labels
+// is the same as iterating through the posting groups and for each group adding all keys.
+func toPostingGroups(ms []*labels.Matcher, indexhdr indexheader.Reader) ([]postingGroup, []labels.Label, error) {
+	var (
+		rawPostingGroups = make([]rawPostingGroup, 0, len(ms))
+		allRequested     = false
+		hasAdds          = false
+	)
+
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, m := range ms {
+		// Each group is separate to tell later what postings are intersecting with what.
+		rawPostingGroups = append(rawPostingGroups, toPostingGroup(m))
+	}
+
+	// First iterate through the non-lazy groups. We can check whether their requested values exist
+	// in the index. After that we can check the lazy groups. The lazy groups require calling LabelValues
+	// which is more expensive, so we leave them for last.
+	sort.Slice(rawPostingGroups, func(i, j int) bool {
+		return (!rawPostingGroups[i].isLazy && rawPostingGroups[j].isLazy) ||
+			(len(rawPostingGroups[i].keys) < len(rawPostingGroups[j].keys))
+	})
+
+	filteredPostingGroups := make([]postingGroup, 0, len(rawPostingGroups))
+	// Next we check whether the posting groups won't select an empty set of postings.
+	// We start with the ones that have a known set of values because it's less expensive to check them in
+	// the index header.
+	numKeys := 0
+	for _, rawGroup := range rawPostingGroups {
+		pg, err := rawGroup.prepare(indexhdr)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "filtering posting group")
+		}
+		filteredPostingGroups = append(filteredPostingGroups, pg)
+
+		// If this group has no keys to work though and is not a subtract group, then it's an empty group.
+		// We can shortcut this, since intersection with empty postings would return no postings.
+		// E.g. `label="non-existent-value"` returns empty group.
+		if !pg.isSubtract && len(pg.keys) == 0 {
+			return nil, nil, nil
+		}
+
+		// If the group is a subtraction group, we must fetch all postings and remove the ones that this matcher selects.
+		allRequested = allRequested || pg.isSubtract
+
+		hasAdds = hasAdds || (!pg.isSubtract && len(pg.keys) > 0)
```

Comment:
```suggestion
		hasAdds = hasAdds || !pg.isSubtract
```

if `len(pg.keys) == 0` then we return above on line 266.